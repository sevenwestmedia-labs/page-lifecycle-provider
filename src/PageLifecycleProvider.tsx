import React from 'react'
import H, { Location } from 'history'
import { Logger, noopLogger } from 'typescript-log'
import { useLocation } from 'react-router'

export interface PageLifecycleEvent<T> {
    type: string
    timeStamp: number
    originator: string
    payload: T
}
export declare interface Properties {
    location: H.Location
    [key: string]: any
}
export interface PageLoadStarted extends PageLifecycleEvent<Properties> {
    type: 'page-load-started'
}

export interface PageLoadFailed
    extends PageLifecycleEvent<Properties & { error: string }> {
    type: 'page-load-failed'
}

export interface PageLoadComplete extends PageLifecycleEvent<Properties> {
    type: 'page-load-complete'
}

export type PageEvent = PageLoadStarted | PageLoadFailed | PageLoadComplete

export interface PageLifecycleProviderRenderProps {
    /** Increments loading count */
    beginLoadingData: () => void
    /** Decrements loading count */
    endLoadingData: () => void
}

export interface PageLifecycleProviderProps {
    children:
        | React.ReactElement<any>
        | ((
              pageProps: PageLifecycleProviderRenderProps,
          ) => React.ReactElement<any>)
    onEvent: (event: PageEvent) => void
    logger?: Logger
}

export interface PageProperties extends PageLifecycleProviderRenderProps {
    currentPageProps: Array<React.MutableRefObject<{}>>
}
export const PagePropertiesContext = React.createContext<PageProperties>(
    undefined as any,
)

export const PageLifecycleProvider: React.FC<PageLifecycleProviderProps> = ({
    children,
    onEvent,
    logger = noopLogger(),
}) => {
    // This will work assuming this component updates before children using the same hook
    const location = useLocation()
    const locationRef = React.useRef(location)
    locationRef.current = location
    const previousLocation = usePrevious(location)
    // This needs to be outside of React's lifecycle so it's immediately consistent
    const loadingDataCount = React.useRef(0)
    // Set to true when location changes, set to false again when update is complete
    const isRouting = React.useRef(true)

    // This needs to be outside React's lifecycle because we want these props to be immediately
    // consistent as it can be updated through context, then we raise events using the latest value
    // If it was stored in React state, it would be out of date
    const contextValue = React.useMemo<PageProperties>(
        () => ({
            currentPageProps: [],
            beginLoadingData,
            endLoadingData,
            currentPageLocation: location,
        }),
        // eslint-disable-next-line react-hooks/exhaustive-deps
        [],
    )

    // While routing our state may be inconsistent
    if (previousLocation && !isRouting.current) {
        // We only care about pathname, not any of the other location info
        if (previousLocation !== location) {
            if (logger) {
                logger.debug(
                    {
                        oldPath: previousLocation.pathname,
                        newPath: location.pathname,
                    },
                    'Path changed',
                )
            }

            isRouting.current = true
        }
    }

    React.useEffect(() => {
        if (!isRouting.current) {
            return
        }

        raisePageLoadStartEvent(
            logger,
            onEvent,
            contextValue.currentPageProps,
            location,
        )
        // No data load triggered, also raise complete event
        if (loadingDataCount.current === 0) {
            raisePageLoadCompleteEvent(
                logger,
                onEvent,
                contextValue.currentPageProps,
                location,
            )
            isRouting.current = false
        }

        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [location, isRouting.current])

    function beginLoadingData() {
        loadingDataCount.current++
        logger.debug(
            { loadingDataCount: loadingDataCount.current },
            'Begin loading data',
        )
    }

    function endLoadingData() {
        loadingDataCount.current--
        logger.debug(
            { loadingDataCount: loadingDataCount.current },
            'End loading data',
        )

        // If pages load data after the page has transitioned to complete, just ignore it
        if (loadingDataCount.current === 0 && isRouting.current) {
            raisePageLoadCompleteEvent(
                logger,
                onEvent,
                contextValue.currentPageProps,
                locationRef.current,
            )
            isRouting.current = false
        }
    }

    if (typeof children === 'function') {
        return (
            <PagePropertiesContext.Provider value={contextValue}>
                {children({
                    beginLoadingData,
                    endLoadingData,
                })}
            </PagePropertiesContext.Provider>
        )
    }

    return (
        <PagePropertiesContext.Provider value={contextValue}>
            {React.Children.only(children)}
        </PagePropertiesContext.Provider>
    )
}

PageLifecycleProvider.displayName = 'PageLifecycleProvider'

function usePrevious<T>(value: T) {
    // The ref object is a generic container whose current property is mutable ...
    // ... and can hold any value, similar to an instance property on a class
    const ref = React.useRef<T>()

    // Store current value in ref
    React.useEffect(() => {
        ref.current = value
    }, [value]) // Only re-run if value changes

    // Return previous value (happens before update in useEffect above)
    return ref.current
}

function raisePageLoadStartEvent(
    logger: Logger,
    onEvent: (event: PageEvent) => void,
    currentPageProps: Array<React.MutableRefObject<{}>>,
    location: Location,
) {
    logger.debug({}, 'Rasing start load event')

    onEvent({
        type: 'page-load-started',
        originator: 'PageEvents',
        payload: {
            ...currentPageProps.reduce(
                (acc, val) => ({ ...acc, ...val.current }),
                {},
            ),
            location,
        },
        timeStamp: new Date().getTime(),
    })
}

function raisePageLoadCompleteEvent(
    logger: Logger,
    onEvent: (event: PageEvent) => void,
    currentPageProps: Array<React.MutableRefObject<{}>>,
    location: Location,
) {
    const reducedProps = currentPageProps.reduce(
        (acc, val) => ({ ...acc, ...val.current }),
        {},
    )

    logger.debug(
        { currentPageProps: reducedProps },
        'Raising page load complete event',
    )

    onEvent({
        type: 'page-load-complete',
        originator: 'PageEvents',
        payload: {
            ...reducedProps,
            location,
        },
        timeStamp: new Date().getTime(),
    })
}
