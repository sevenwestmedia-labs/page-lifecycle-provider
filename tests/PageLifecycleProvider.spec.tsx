/**
 * @jest-environment jsdom
 */

window.requestAnimationFrame = (callback: any) => {
    return setTimeout(callback, 0) as any
}

import H from 'history'
import React from 'react'
import { act } from 'react-dom/test-utils'
import Adapter from 'enzyme-adapter-react-16'
import { MemoryRouter, Route } from 'react-router-dom'
import { mount, configure } from 'enzyme'

import { PromiseCompletionSource } from 'promise-completion-source'
import {
    PageEvent,
    PageLifecycleProvider,
    PagePropertiesContext,
} from '../src/PageLifecycleProvider'
import { PageProps } from '../src/PageProps'

configure({ adapter: new Adapter() })

interface TestData {
    bar: string
}

const createTestComponents = (asyncCompletion = true) => {
    const promiseCompletionSource = new PromiseCompletionSource<TestData>()
    let loadTriggered = false
    const TestPage: React.FC<{ extraProps?: object }> = ({ extraProps }) => {
        return (
            <PageProps pageProperties={extraProps}>
                {(pageProps) => {
                    // This emulates a component under the page starting to load data
                    // then completing once the promise completion source completes
                    if (!loadTriggered) {
                        pageProps.beginLoadingData()

                        loadTriggered = true
                        if (asyncCompletion) {
                            promiseCompletionSource.promise.then(() =>
                                pageProps.endLoadingData(),
                            )
                        } else {
                            pageProps.endLoadingData()
                        }
                    }

                    return null
                }}
            </PageProps>
        )
    }

    return {
        promiseCompletionSource,
        TestPage,
    }
}

describe('PageLifecycleProvider', () => {
    it('raises loading event after render', () => {
        const testComponents = createTestComponents()
        const pageEvents: PageEvent[] = []

        const wrapper = mount(
            <MemoryRouter initialEntries={['/']} initialIndex={0}>
                <PageLifecycleProvider
                    onEvent={(event) => pageEvents.push(event)}
                >
                    <testComponents.TestPage />
                </PageLifecycleProvider>
            </MemoryRouter>,
        )

        wrapper.find(testComponents.TestPage)

        expect(
            pageEvents.map((e) => {
                e.timeStamp = 0
                if (e.payload && e.payload.location) {
                    e.payload.location.key = '...'
                }
                return e
            }),
        ).toMatchSnapshot()
    })

    it('raises start event before complete ', () => {
        const testComponents = createTestComponents(false)
        const pageEvents: PageEvent[] = []

        const wrapper = mount(
            <MemoryRouter initialEntries={['/']} initialIndex={0}>
                <PageLifecycleProvider
                    onEvent={(event) => pageEvents.push(event)}
                >
                    <testComponents.TestPage />
                </PageLifecycleProvider>
            </MemoryRouter>,
        )

        wrapper.find(testComponents.TestPage)

        expect(
            pageEvents.map((e) => {
                e.timeStamp = 0
                if (e.payload && e.payload.location) {
                    e.payload.location.key = '...'
                }
                return e
            }),
        ).toMatchSnapshot()
    })

    it('can pass extra data to pages', async () => {
        const testComponents = createTestComponents()
        let history: H.History | undefined
        const pageEvents: PageEvent[] = []
        const extraPropsLookup: { [path: string]: object } = {
            '/': { home: 'isHome' },
            '/foo': { foo: 'isFoo' },
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/', '/foo']} initialIndex={0}>
                <PageLifecycleProvider
                    onEvent={(event) => pageEvents.push(event)}
                >
                    <Route
                        render={(props) => {
                            history = props.history
                            return (
                                <div>
                                    {props.location.pathname === '/' && (
                                        <PageProps
                                            pageProperties={{
                                                pageExtra:
                                                    'Some extra page data',
                                            }}
                                        />
                                    )}
                                    <testComponents.TestPage
                                        extraProps={
                                            extraPropsLookup[
                                                props.location.pathname
                                            ]
                                        }
                                    />
                                </div>
                            )
                        }}
                    />
                </PageLifecycleProvider>
            </MemoryRouter>,
        )

        if (!history) {
            throw new Error('History not defined')
        }

        act(() => {
            testComponents.promiseCompletionSource.resolve({ bar: 'test' })
        })
        await new Promise((resolve) => setTimeout(() => resolve()))
        testComponents.promiseCompletionSource = new PromiseCompletionSource()

        act(() => {
            history!.push('/foo')

            testComponents.promiseCompletionSource.resolve({ bar: 'page2' })
        })
        await new Promise((resolve) => setTimeout(() => resolve()))

        wrapper.update().find(testComponents.TestPage)
        expect(
            pageEvents.map((e) => {
                e.timeStamp = 0
                if (e.payload && e.payload.location) {
                    e.payload.location.key = '...'
                }
                return e
            }),
        ).toMatchSnapshot()
    })

    it('raises completed event after data loaded and content re-rendered', async () => {
        const testComponents = createTestComponents()
        const pageEvents: PageEvent[] = []

        const wrapper = mount(
            <MemoryRouter initialEntries={['/']} initialIndex={0}>
                <PageLifecycleProvider
                    onEvent={(event) => pageEvents.push(event)}
                >
                    <testComponents.TestPage />
                </PageLifecycleProvider>
            </MemoryRouter>,
        )

        act(() => {
            testComponents.promiseCompletionSource.resolve({
                bar: 'test',
            })
        })

        expect(pageEvents.length).toBe(1)
        await new Promise((resolve) => setTimeout(() => resolve()))
        wrapper.update().find(testComponents.TestPage)

        expect(
            pageEvents.map((e) => {
                e.timeStamp = 0
                if (e.payload && e.payload.location) {
                    e.payload.location.key = '...'
                }
                return e
            }),
        ).toMatchSnapshot()
    })

    it('raises completed event after lazy loaded page has completed load data', async () => {
        const testComponents = createTestComponents()
        const pageEvents: PageEvent[] = []

        const wrapper = mount(
            <MemoryRouter initialEntries={['/']} initialIndex={0}>
                <PageLifecycleProvider
                    onEvent={(event) => pageEvents.push(event)}
                >
                    <testComponents.TestPage />
                </PageLifecycleProvider>
            </MemoryRouter>,
        )

        wrapper.find(testComponents.TestPage)

        expect(
            pageEvents.map((e) => {
                e.timeStamp = 0
                if (e.payload && e.payload.location) {
                    e.payload.location.key = '...'
                }
                return e
            }),
        ).toMatchSnapshot()

        await new Promise((resolve) => setTimeout(() => resolve()))

        act(() => {
            testComponents.promiseCompletionSource.resolve({
                bar: 'test',
            })
        })
        // We should not have received the completed event yet
        expect(pageEvents.length).toBe(1)

        // Let the completion propagate
        await new Promise((resolve) => setTimeout(() => resolve()))
        expect(
            pageEvents.map((e) => {
                e.timeStamp = 0
                if (e.payload && e.payload.location) {
                    e.payload.location.key = '...'
                }
                return e
            }),
        ).toMatchSnapshot()
    })

    it('raises events when page navigated', async () => {
        const testComponents = createTestComponents()
        let history: H.History | undefined
        const pageEvents: PageEvent[] = []

        const wrapper = mount(
            <MemoryRouter initialEntries={['/', '/foo']} initialIndex={0}>
                <PageLifecycleProvider
                    onEvent={(event) => pageEvents.push(event)}
                >
                    <Route
                        render={(props) => {
                            history = props.history
                            return <testComponents.TestPage />
                        }}
                    />
                </PageLifecycleProvider>
            </MemoryRouter>,
        )

        if (!history) {
            throw new Error('History not defined')
        }

        act(() => {
            testComponents.promiseCompletionSource.resolve({ bar: 'test' })
        })
        await new Promise((resolve) => setTimeout(() => resolve()))
        testComponents.promiseCompletionSource = new PromiseCompletionSource()

        act(() => {
            history!.push('/foo')
            testComponents.promiseCompletionSource.resolve({ bar: 'page2' })
        })
        await new Promise((resolve) => setTimeout(() => resolve()))
        wrapper.update().find(testComponents.TestPage)
        expect(
            pageEvents.map((e) => {
                e.timeStamp = 0
                if (e.payload && e.payload.location) {
                    e.payload.location.key = '...'
                }
                return e
            }),
        ).toMatchSnapshot()
    })

    it('Does not raise events twice for a single route', async () => {
        const testComponents = createTestComponents()
        let history: H.History | undefined
        const pageEvents: PageEvent[] = []
        let triggerDataLoad: () => void | undefined
        let triggerDataLoadComplete: () => void | undefined

        const wrapper = mount(
            <MemoryRouter initialEntries={['/', '/foo']} initialIndex={0}>
                <PageLifecycleProvider
                    onEvent={(event) => pageEvents.push(event)}
                >
                    <Route
                        render={(props) => {
                            history = props.history
                            return (
                                <div>
                                    <testComponents.TestPage />
                                    <PageProps>
                                        {(pageProps) => {
                                            triggerDataLoad =
                                                pageProps.beginLoadingData
                                            triggerDataLoadComplete =
                                                pageProps.endLoadingData

                                            return null
                                        }}
                                    </PageProps>
                                </div>
                            )
                        }}
                    />
                </PageLifecycleProvider>
            </MemoryRouter>,
        )

        if (!history) {
            throw new Error('History not defined')
        }

        act(() => {
            testComponents.promiseCompletionSource.resolve({ bar: 'test' })
        })
        await new Promise((resolve) => setTimeout(() => resolve()))

        act(() => {
            triggerDataLoad!()
            triggerDataLoadComplete!()
        })

        expect(
            pageEvents.map((e) => {
                e.timeStamp = 0
                if (e.payload && e.payload.location) {
                    e.payload.location.key = '...'
                }
                return e
            }),
        ).toMatchSnapshot()
    })
})
