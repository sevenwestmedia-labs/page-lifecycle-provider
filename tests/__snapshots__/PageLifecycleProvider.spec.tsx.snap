// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageLifecycleProvider Does not raise events twice for a single route 1`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
]
`;

exports[`PageLifecycleProvider can pass extra data to pages 1`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "home": "isHome",
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "pageExtra": "Some extra page data",
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "home": "isHome",
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "pageExtra": "Some extra page data",
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "foo": "isFoo",
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/foo",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "foo": "isFoo",
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/foo",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
]
`;

exports[`PageLifecycleProvider raises completed event after data loaded and content re-rendered 1`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
]
`;

exports[`PageLifecycleProvider raises completed event after lazy loaded page has completed load data 1`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
]
`;

exports[`PageLifecycleProvider raises completed event after lazy loaded page has completed load data 2`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
]
`;

exports[`PageLifecycleProvider raises events when page navigated 1`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/foo",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/foo",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-complete",
  },
]
`;

exports[`PageLifecycleProvider raises loading event after render 1`] = `
Array [
  Object {
    "originator": "PageEvents",
    "payload": Object {
      "location": Object {
        "hash": "",
        "key": "...",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
    },
    "timeStamp": 0,
    "type": "page-load-started",
  },
]
`;
